File: Source/frontend/backend/routes/alerts.js
import { Router } from 'express';
import { alertStatus, alertSeverityDisplay } from '../utils/helpers.js';

const router = Router();

/**
 * @swagger
 * /api/alerts:
 *   get:
 *     tags: [Alerts]
 *     summary: Get alerts with filtering
 *     description: Retrieve alerts with optional severity filtering
 *     parameters:
 *       - in: query
 *         name: severity
 *         schema:
 *           type: string
 *           enum: [low, medium, high, critical]
 *         description: Alert severity filter
 *       - in: query
 *         name: limit
 *         schema:
 *           type: integer
 *           default: 50
 *         description: Maximum number of alerts to return
 *     responses:
 *       200:
 *         description: Successful response
 *         content:
 *           application/json:
 *             schema:
 *               type: array
 *               items:
 *                 type: object
 *                 properties:
 *                   id:
 *                     type: string
 *                   zoneId:
 *                     type: string
 *                   severity:
 *                     type: string
 *                   type:
 *                     type: string
 *                   title:
 *                     type: string
 *                   description:
 *                     type: string
 *                   status:
 *                     type: string
 *                   timestamp:
 *                     type: string
 *                     format: date-time
 */
router.get('/', async (req, res) => {
  try {
    const { severity, limit = 50 } = req.query;

    let query = `
      SELECT
        a.*,
        z.name as zone_name,
        z.code as zone_code
      FROM alerts a
      JOIN zones z ON a.zone_id = z.id
      WHERE a.created_at >= NOW() - INTERVAL '72 hours'
    `;

    const params = [];
    let paramIndex = 1;

    if (severity) {
      query += ` AND a.severity = $${paramIndex++}`;
      params.push(severity);
    }

    query += ` ORDER BY a.created_at DESC LIMIT $${paramIndex++}`;
    params.push(parseInt(limit));

    const result = await req.pool.query(query);

    const alerts = result.rows.map(row => ({
      id: row.metadata?.externalId || row.id,
      zoneId: row.zone_code,
      severity: alertSeverityDisplay(row.severity),
      type: row.alert_type === 'crew' ? 'Crew' : 'System',
      crewId: row.metadata?.crewId,
      title: row.title,
      description: row.message,
      eta: row.metadata?.eta || null,
      status: row.metadata?.status ?? alertStatus(row),
      timestamp: row.created_at instanceof Date ? row.created_at.toISOString() : row.created_at,
    }));

    res.json(alerts);
  } catch (error) {
    console.error('Database error:', error);
    res.status(500).json({
      error: 'Failed to fetch alerts',
      details: process.env.NODE_ENV === 'development' ? error.message : undefined
    });
  }
});

/**
 * @swagger
 * /api/alerts/{id}/ack:
 *   post:
 *     tags: [Alerts]
 *     summary: Acknowledge alert
 *     description: Mark an alert as acknowledged
 *     parameters:
 *       - in: path
 *         name: id
 *         required: true
 *         schema:
 *           type: string
 *         description: Alert ID
 *     requestBody:
 *       required: true
 *       content:
 *         application/json:
 *           schema:
 *             type: object
 *             properties:
 *               acknowledgedBy:
 *                 type: string
 *                 description: Name of person acknowledging
 *                 example: John Doe
 *     responses:
 *       200:
 *         description: Alert acknowledged successfully
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                 success:
 *                   type: boolean
 *                   example: true
 *                 message:
 *                   type: string
 *                   example: Alert acknowledged successfully
 */
router.post('/:id/ack', async (req, res) => {
  try {
    const { id } = req.params;
    const { acknowledgedBy } = req.body;

    const query = `
      UPDATE alerts
      SET acknowledged = true,
          acknowledged_at = NOW(),
          acknowledged_by = COALESCE($1, acknowledged_by),
          metadata = jsonb_set(
            metadata || '{}',
            '{status}',
            '"acknowledged"',
            true
          )
      WHERE id = $2
      RETURNING *
    `;

    const result = await req.pool.query(query, [acknowledgedBy, id]);

    if (result.rows.length === 0) {
      return res.status(404).json({
        error: 'Alert not found',
        details: `Alert with ID ${id} does not exist`
      });
    }

    res.json({
      success: true,
      message: 'Alert acknowledged successfully',
      alert: {
        id: result.rows[0].id,
        acknowledged: true,
        acknowledgedAt: result.rows[0].acknowledged_at,
        acknowledgedBy: result.rows[0].acknowledged_by
      }
    });
  } catch (error) {
    console.error('Database error:', error);
    res.status(500).json({
      error: 'Failed to acknowledge alert',
      details: process.env.NODE_ENV === 'development' ? error.message : undefined
    });
  }
});

export default router;

File: Source/frontend/python_backend/app/routers/alerts.py
"""Alert endpoints."""
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import text
from typing import Optional, List
from datetime import datetime

from ..database import get_db
from ..schemas import AlertResponse, AlertAckRequest, AlertAckResponse
from ..utils import get_alert_status, get_alert_severity_display

router = APIRouter(prefix="/api", tags=["Alerts"])


@router.get("/alerts", response_model=List[AlertResponse])
async def get_alerts(
    severity: Optional[str] = Query(None, description="Filter by severity level"),
    limit: int = Query(50, ge=1, le=500, description="Maximum number of alerts"),
    db: AsyncSession = Depends(get_db)
):
    """Get alerts with filtering options."""
    try:
        params = {"limit": limit}
        severity_clause = ""
        
        if severity:
            severity_clause = "AND a.severity = :severity"
            params["severity"] = severity
        
        query = text(f"""
            SELECT
                a.*,
                z.name as zone_name,
                z.code as zone_code
            FROM alerts a
            JOIN zones z ON a.zone_id = z.id
            WHERE a.created_at >= NOW() - INTERVAL '72 hours'
            {severity_clause}
            ORDER BY a.created_at DESC
            LIMIT :limit
        """)
        
        result = await db.execute(query, params)
        rows = result.fetchall()
        
        alerts = []
        for row in rows:
            metadata = getattr(row, "meta", None) or {}
            row_dict = {
                'resolved': row.resolved,
                'acknowledged': row.acknowledged,
            }
            
            timestamp = row.created_at
            if isinstance(timestamp, datetime):
                timestamp = timestamp.isoformat()
            
            alerts.append({
                "id": metadata.get('externalId', str(row.id)),
                "zoneId": row.zone_code,
                "severity": get_alert_severity_display(row.severity),
                "type": "Crew" if row.alert_type == "crew" else "System",
                "crewId": metadata.get('crewId'),
                "title": row.title,
                "description": row.message,
                "eta": metadata.get('eta'),
                "status": metadata.get('status') or get_alert_status(row_dict),
                "timestamp": timestamp,
            })
        
        return alerts
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to fetch alerts: {str(e)}")


@router.post("/alerts/{id}/ack", response_model=AlertAckResponse)
async def acknowledge_alert(
    id: str,
    request: AlertAckRequest,
    db: AsyncSession = Depends(get_db)
):
    """Mark an alert as acknowledged."""
    try:
        query = text("""
            UPDATE alerts
            SET acknowledged = true,
                acknowledged_by = :acknowledged_by,
                acknowledged_at = NOW()
            WHERE id::text = :id OR metadata->>'externalId' = :id
            RETURNING *
        """)
        
        result = await db.execute(query, {
            "acknowledged_by": request.acknowledgedBy or "system",
            "id": id
        })
        row = result.fetchone()
        
        if not row:
            raise HTTPException(status_code=404, detail="Alert not found")
        
        await db.commit()
        
        return {
            "success": True,
            "message": "Alert acknowledged successfully",
            "alert": {"id": str(row.id)}
        }
    except HTTPException:
        raise
    except Exception as e:
        await db.rollback()
        raise HTTPException(status_code=500, detail=f"Failed to acknowledge alert: {str(e)}")
